cmd/secret-collection-manager/main.go

```diff
@@ -0,0 +1,415 @@
+	flag.StringVar(&o.vaultToken, "vault-token", "jpuxZFWWFW7vM882GGX2aWOE", "The privileged token to use when communicating with vault, must be able to CRUD policies")
+	flag.Parse()
+	return o
+}
+
+func main() {
+	logrusutil.ComponentInit()
+	o := parseOptions()
+
+	privilegedVaultClient, err := vaultclient.New(o.vaultAddr, o.vaultToken)
+	if err != nil {
+		logrus.WithError(err).Fatal("Failed to construct vault client")
+	}
+
+	if err := server(privilegedVaultClient, o.kvStorePrefix, o.listenAddr).ListenAndServe(); err != nil && err != http.ErrServerClosed {
+		logrus.WithError(err).Fatal("failed to listen and serve")
+	}
+
+}
+
+func server(privilegedVaultClient *vaultclient.VaultClient, kvStorePrefix, listenAddr string) *http.Server {
+	manager := &secretCollectionManager{
+		privilegedVaultClient: privilegedVaultClient,
+		kvStorePrefix:         kvStorePrefix,
+	}
+
+	return &http.Server{Addr: listenAddr, Handler: manager.mux()}
+}
+
+func userWrapper(upstream func(user string, w http.ResponseWriter, r *http.Request, params httprouter.Params)) httprouter.Handle {
+	return func(w http.ResponseWriter, r *http.Request, params httprouter.Params) {
+		user := strings.Split(r.Header.Get("X-Forwarded-Email"), "@")[0]
+		if user == "" {
+			http.Error(w, "No user passed", 400)
+			logrus.WithField("X-Forwarded-Email", r.Header.Get("X-Forwarded-Email")).Error("Got request with empty user")
+			return
+		}
+
+		upstream(user, w, r, params)
+	}
+}
+
+func loggingWrapper(upstream func(l *logrus.Entry, user string, w http.ResponseWriter, r *http.Request, params httprouter.Params)) func(string, http.ResponseWriter, *http.Request, httprouter.Params) {
+	return func(user string, w http.ResponseWriter, r *http.Request, params httprouter.Params) {
+		upstream(logrus.WithField("UID", uuid.NewV1().String()).WithField("user", user), user, w, r, params)
+	}
+}
+
+type secretCollectionManager struct {
+	privilegedVaultClient *vaultclient.VaultClient
+	kvStorePrefix         string
+	groupCache            idNameCache
+	userCache             idNameCache
+}
+
+// idNameCache allows to get the id or the name, using
+// the opposing attribute. It assumes that the relationship
+// is immmutable.
+type idNameCache struct {
+	lock  sync.RWMutex
+	names map[string]string
+	ids   map[string]string
+}
+
+func (c *idNameCache) byID(id string) string {
+	c.lock.RLock()
+	defer c.lock.RUnlock()
+	return c.ids[id]
+}
+
+func (c *idNameCache) byName(name string) string {
+	c.lock.RLock()
+	defer c.lock.RUnlock()
+	return c.names[name]
+}
+
+func (c *idNameCache) set(name, id string) {
+	c.lock.Lock()
+	defer c.lock.Unlock()
+	if c.names == nil {
+		c.names = map[string]string{}
+	}
+	if c.ids == nil {
+		c.ids = map[string]string{}
+	}
+	c.names[name] = id
+	c.ids[id] = name
+}
+
+func (m *secretCollectionManager) mux() *httprouter.Router {
+	router := httprouter.New()
+	router.GET("/healthz", healthHandler)
```

Comment:
We have health & readiness lib already from Prow, prefer that for consistency of code, ports, etc

Reply:
It doesn't really do anything different other than hardcoding a different port